internal/bminventory/inventory.go

```diff
@@ -735,34 +738,100 @@ func (b *bareMetalInventory) createAndUploadDay2NodeIgnition(ctx context.Context
-		ignitionEndpointUrl = url.String()
+		caCertPointer := cluster.IgnitionEndpoint.CaCertificate
+		if caCertPointer != nil {
+			return url.String(), *caCertPointer, nil
+		}
+		return url.String(), "", nil
 	}
+	return "", "", nil
+}
 
-	var caCert *string = nil
-	if cluster.IgnitionEndpoint != nil {
-		caCert = cluster.IgnitionEndpoint.CaCertificate
+func (b *bareMetalInventory) getIgnitionURL(ctx context.Context, cluster *common.Cluster, address string, host *models.Host) (string, *string) {
+	mcpName := host.MachineConfigPoolName
+	if customIgnitionUrl, customCACert, err := b.clusterHasCustomIgnitionURL(cluster, host); err == nil {
+		if err == nil && customIgnitionUrl != "" {
+			return customIgnitionUrl, &customCACert
+		}
+	}
+	k8sClient, kubeconfigPath, err := b.GetKubeconfigForCluster(ctx, cluster)
+	if err == nil {
+		defer os.Remove(kubeconfigPath)
+		if ignitionUrlFromCluster, caCertFromCluster, err := b.fetchMCSCertificateFromCluster(ctx, k8sClient, address, mcpName); err == nil {
+			return ignitionUrlFromCluster, &caCertFromCluster
+		}
 	}
+	return fmt.Sprintf("http://%s:22624/config/%s", address, mcpName), nil
+}
 
-	fullIgnition, err := b.IgnitionBuilder.FormatSecondDayWorkerIgnitionFile(ignitionEndpointUrl, caCert, ignitionEndpointToken, host)
+func (b *bareMetalInventory) GetKubeconfigForCluster(ctx context.Context, cluster *common.Cluster) (k8sclient.K8SClient, string, error) {
+	respBody, _, err := b.V2DownloadClusterCredentialsInternal(ctx, installer.V2DownloadClusterCredentialsParams{
+		ClusterID: *cluster.ID,
+		FileName:  constants.Kubeconfig,
+	})
 	if err != nil {
-		return errors.Wrapf(err, "Failed to create ignition string for cluster %s, host %s", cluster.ID, host.ID)
+		return nil, "", err
+	}
+	respBytes, err := ioutil.ReadAll(respBody)
+	if err != nil || len(respBytes) == 0 {
+		return nil, "", err
+	}
+	file, err := ioutil.TempFile("", "kubeconfig")
+	if err != nil {
+		return nil, "", err
+	}
+	err = os.WriteFile(file.Name(), respBytes, 0644)
+	if err != nil {
+		return nil, "", err
+	}
+	k8sClusterClient, err := b.k8sClient.NewFromKubeconfig(file.Name())
+	if err != nil {
+		return nil, "", err
 	}
+	return k8sClusterClient, file.Name(), err
+}
 
-	fileName := fmt.Sprintf("%s/worker-%s.ign", cluster.ID, host.ID)
-	log.Infof("Uploading ignition file <%s>", fileName)
-	err = b.objectHandler.Upload(ctx, fullIgnition, fileName)
+func (b *bareMetalInventory) fetchMCSCertificateFromCluster(ctx context.Context, k8sClient k8sclient.K8SClient, address, machineConfigPoolName string) (string, string, error) {
+	mcsSecret, err := k8sClient.GetSecret("openshift-machine-config-operator", "machine-config-server-tls")
 	if err != nil {
-		return errors.Errorf("Failed to upload worker ignition for cluster %s", cluster.ID)
+		return "", "", err
 	}
-	return nil
+	caCertBytesBase64, ok := mcsSecret.Data["tls.crt"]
+	if !ok {
+		return "", "", errors.Errorf("MCS secret doesn't have expected tls key")
```

Comment:
Why does this function construct the ignition URL? The names seem to imply it should be a responsibility of the caller